base/cvd/cuttlefish/common/libs/utils/archive.cpp

```diff
@@ -115,9 +117,72 @@ Result<std::string> ExtractImage(const std::string& archive_filepath,
   return {result.front()};
 }
 
+namespace {
+
+Result<std::string> hardlinkRecursively(const std::string& src_dir,
+                                        const std::string& dst_dir,
+                                        std::vector<std::string>& files) {
+  DIR* dir = opendir(src_dir.c_str());
```

Comment:
use the `WalkDirectory` function in `base/cvd/cuttlefish/common/libs/util/files.h`. If you need more control on the order in which nodes are visited you can use `DirectoryContents` and `DirectoryExists` from the same header file.